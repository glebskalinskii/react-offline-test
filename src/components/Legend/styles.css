.legend {
  width: 500px;
  font-size: 24px;
}

.legend-item {
  display: flex;
  padding: 9px 0;
  color: var(--color-white);
}

.has-active-item .legend-item {
  opacity: .4;
}

.legend-item.active {
  opacity: 1;
}

.legend-item-key {
  display: inline-block;
  width: 24px;
  height: 24px;
  border-radius: 50%;
  margin-right: 9px;
}

.legend-item-fuel {
  flex-grow: 1;
}
