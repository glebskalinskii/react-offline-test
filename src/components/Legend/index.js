import React from 'react';
import cx from 'classnames';

import './styles.css';

const Legend = ({ data, activeIndex, setActiveIndex }) => {
  const listClassname = cx('legend', {'has-active-item': activeIndex !== null});

  return <ul className={listClassname}>
    {data.map((energyUnit, i) => <li
      className={`legend-item ${activeIndex === i ? 'active' : ''}`}
      key={energyUnit.fuel}
      onMouseEnter={() => setActiveIndex(i)}
      onMouseLeave={() => setActiveIndex(null)}
    >
      <span className="legend-item-key" style={{background: energyUnit.colorKey}}></span>
      <span className="legend-item-fuel">{energyUnit.fuel}</span>
      <span>{`${energyUnit.perc}%`}</span>
    </li>)}
  </ul>
}

export default Legend;
