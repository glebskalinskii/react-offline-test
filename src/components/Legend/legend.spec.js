import React from 'react';
import { shallow } from 'enzyme';
import Legend from './index';
import testData from '../../../example_api_response.json';

describe('Legend', () => {
  test('should render and match snapshot', () => {
    const wrapper = shallow(<Legend
        data={testData.data.generationmix}
        activeIndex={null}
        setActiveIndex={() => {}}
      />);

    expect(wrapper).toMatchSnapshot();
  });

  test('should highlight active item and match snapshot', () => {
    const wrapper = shallow(<Legend
        data={testData.data.generationmix}
        activeIndex={0}
        setActiveIndex={() => {}}
      />);

    expect(wrapper).toMatchSnapshot();
  });

  test('should call mouseenter callback', () => {
    const setActiveIndexMock = jest.fn();

    const wrapper = shallow(<Legend
        data={testData.data.generationmix}
        activeIndex={null}
        setActiveIndex={setActiveIndexMock}
      />);
    const firstItem = wrapper.find('li').first();

    firstItem.simulate('mouseenter');

    expect(setActiveIndexMock.mock.calls.length).toBe(1);
    expect(setActiveIndexMock.mock.calls[0][0]).toBe(0);
  });

  test('should call mouseleave callback', () => {
    const setActiveIndexMock = jest.fn();

    const wrapper = shallow(<Legend
        data={testData.data.generationmix}
        activeIndex={null}
        setActiveIndex={setActiveIndexMock}
      />);
    const firstItem = wrapper.find('li').first();

    firstItem.simulate('mouseleave');

    expect(setActiveIndexMock.mock.calls.length).toBe(1);
    expect(setActiveIndexMock.mock.calls[0][0]).toBe(null);
  });
});

