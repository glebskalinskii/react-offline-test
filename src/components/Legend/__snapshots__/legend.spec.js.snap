// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Legend should highlight active item and match snapshot 1`] = `
<ul
  className="legend has-active-item"
>
  <li
    className="legend-item active"
    key="biomass"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <span
      className="legend-item-key"
      style={
        Object {
          "background": undefined,
        }
      }
    />
    <span
      className="legend-item-fuel"
    >
      biomass
    </span>
    <span>
      4.8%
    </span>
  </li>
  <li
    className="legend-item "
    key="coal"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <span
      className="legend-item-key"
      style={
        Object {
          "background": undefined,
        }
      }
    />
    <span
      className="legend-item-fuel"
    >
      coal
    </span>
    <span>
      2.5%
    </span>
  </li>
  <li
    className="legend-item "
    key="imports"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <span
      className="legend-item-key"
      style={
        Object {
          "background": undefined,
        }
      }
    />
    <span
      className="legend-item-fuel"
    >
      imports
    </span>
    <span>
      8.7%
    </span>
  </li>
  <li
    className="legend-item "
    key="gas"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <span
      className="legend-item-key"
      style={
        Object {
          "background": undefined,
        }
      }
    />
    <span
      className="legend-item-fuel"
    >
      gas
    </span>
    <span>
      46.5%
    </span>
  </li>
  <li
    className="legend-item "
    key="nuclear"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <span
      className="legend-item-key"
      style={
        Object {
          "background": undefined,
        }
      }
    />
    <span
      className="legend-item-fuel"
    >
      nuclear
    </span>
    <span>
      16.1%
    </span>
  </li>
  <li
    className="legend-item "
    key="other"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <span
      className="legend-item-key"
      style={
        Object {
          "background": undefined,
        }
      }
    />
    <span
      className="legend-item-fuel"
    >
      other
    </span>
    <span>
      0.3%
    </span>
  </li>
  <li
    className="legend-item "
    key="hydro"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <span
      className="legend-item-key"
      style={
        Object {
          "background": undefined,
        }
      }
    />
    <span
      className="legend-item-fuel"
    >
      hydro
    </span>
    <span>
      0.9%
    </span>
  </li>
  <li
    className="legend-item "
    key="solar"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <span
      className="legend-item-key"
      style={
        Object {
          "background": undefined,
        }
      }
    />
    <span
      className="legend-item-fuel"
    >
      solar
    </span>
    <span>
      14.6%
    </span>
  </li>
  <li
    className="legend-item "
    key="wind"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <span
      className="legend-item-key"
      style={
        Object {
          "background": undefined,
        }
      }
    />
    <span
      className="legend-item-fuel"
    >
      wind
    </span>
    <span>
      5.6%
    </span>
  </li>
</ul>
`;

exports[`Legend should render and match snapshot 1`] = `
<ul
  className="legend"
>
  <li
    className="legend-item "
    key="biomass"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <span
      className="legend-item-key"
      style={
        Object {
          "background": undefined,
        }
      }
    />
    <span
      className="legend-item-fuel"
    >
      biomass
    </span>
    <span>
      4.8%
    </span>
  </li>
  <li
    className="legend-item "
    key="coal"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <span
      className="legend-item-key"
      style={
        Object {
          "background": undefined,
        }
      }
    />
    <span
      className="legend-item-fuel"
    >
      coal
    </span>
    <span>
      2.5%
    </span>
  </li>
  <li
    className="legend-item "
    key="imports"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <span
      className="legend-item-key"
      style={
        Object {
          "background": undefined,
        }
      }
    />
    <span
      className="legend-item-fuel"
    >
      imports
    </span>
    <span>
      8.7%
    </span>
  </li>
  <li
    className="legend-item "
    key="gas"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <span
      className="legend-item-key"
      style={
        Object {
          "background": undefined,
        }
      }
    />
    <span
      className="legend-item-fuel"
    >
      gas
    </span>
    <span>
      46.5%
    </span>
  </li>
  <li
    className="legend-item "
    key="nuclear"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <span
      className="legend-item-key"
      style={
        Object {
          "background": undefined,
        }
      }
    />
    <span
      className="legend-item-fuel"
    >
      nuclear
    </span>
    <span>
      16.1%
    </span>
  </li>
  <li
    className="legend-item "
    key="other"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <span
      className="legend-item-key"
      style={
        Object {
          "background": undefined,
        }
      }
    />
    <span
      className="legend-item-fuel"
    >
      other
    </span>
    <span>
      0.3%
    </span>
  </li>
  <li
    className="legend-item "
    key="hydro"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <span
      className="legend-item-key"
      style={
        Object {
          "background": undefined,
        }
      }
    />
    <span
      className="legend-item-fuel"
    >
      hydro
    </span>
    <span>
      0.9%
    </span>
  </li>
  <li
    className="legend-item "
    key="solar"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <span
      className="legend-item-key"
      style={
        Object {
          "background": undefined,
        }
      }
    />
    <span
      className="legend-item-fuel"
    >
      solar
    </span>
    <span>
      14.6%
    </span>
  </li>
  <li
    className="legend-item "
    key="wind"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <span
      className="legend-item-key"
      style={
        Object {
          "background": undefined,
        }
      }
    />
    <span
      className="legend-item-fuel"
    >
      wind
    </span>
    <span>
      5.6%
    </span>
  </li>
</ul>
`;
